# LSTM + Random Forest Hybrid (Hybrid 1) — CORRECTED PIPELINE

**This notebook supersedes the previous Colab-based run.**

Change log (2026-04-24):
- Loads the post-alignment-fix CSVs `fraudTrain_engineered_with_ids.csv` / `fraudTest_engineered_with_ids.csv` (composite-key merge, validated one-to-one — see `AUDIT_WIN_NARRATIVE.md`).
- Removes the broken positional `.values` reattachment of `cc_num` from the raw CSV.
- Removes Google Drive mount / Colab paths.
- Adds fixed random seeds for reproducibility.

**Corrected LSTM+RF F1 = 0.7892** (from `run_gap_experiments.py`, seed=42, recorded in `verified_metrics.json["gap_experiments"]["LSTM_reproduced_baseline"]`).

The historical F1 = 0.4747 from the original Colab run was produced by the broken positional-attach pipeline and is preserved only in `FYP_Hybrid_Model_BROKEN.ipynb` for audit. It should not be cited as the current Hybrid 1 result.


In [1]:
# Reproducibility seeds
import random
import numpy as np

SEED = 42
random.seed(SEED)
np.random.seed(SEED)

try:
    import torch
    torch.manual_seed(SEED)
    if torch.cuda.is_available():
        torch.cuda.manual_seed_all(SEED)
except ImportError:
    pass

try:
    import tensorflow as tf
    tf.random.set_seed(SEED)
except ImportError:
    pass


In [2]:
import pandas as pd
import numpy as np

# Post-fix CSVs: cc_num and trans_date_trans_time come from a validated
# composite-key one-to-one merge (see AUDIT_WIN_NARRATIVE.md).
df_train = pd.read_csv('../data/engineered/fraudTrain_engineered_with_ids.csv')
df_test  = pd.read_csv('../data/engineered/fraudTest_engineered_with_ids.csv')

target = 'is_fraud'
drop_cols = ['is_fraud', 'unix_time']  # features exclude these; cc_num/trans_date_trans_time
                                        # stay in X_train/X_test for sequence grouping only

X_train = df_train.drop(columns=drop_cols)
y_train = df_train[target]
X_test  = df_test.drop(columns=drop_cols)
y_test  = df_test[target]

print(f"Train: {X_train.shape} | Fraud: {y_train.sum()}")
print(f"Test:  {X_test.shape} | Fraud: {y_test.sum()}")
print(f"Unique cards in train: {X_train['cc_num'].nunique()}")
print(f"Unique cards in test:  {X_test['cc_num'].nunique()}")


Train: (1296675, 16) | Fraud: 7506
Test:  (555719, 16) | Fraud: 2145
Unique cards in train: 983
Unique cards in test:  924


## Note on cardholder IDs

The original notebook attached `cc_num` and `trans_date_trans_time` by positional `.values` from the raw CSV, assuming row `i` of the engineered file corresponded to row `i` of the raw file. It did not — the EDA notebook sorts rows by `(cc_num, trans_datetime)` during velocity computation, so positional alignment was broken (0 / 1,296,675 rows matched — see `verify_alignment_bug.py`).

In this corrected notebook, `cc_num` and `trans_date_trans_time` are already present in the engineered CSV via a validated `(unix_time, amt, city_pop)` composite-key merge (`fix_lstm_alignment.py`, `validate="one_to_one"`, plus a 20-row round-trip sanity check). The positional reattachment step is therefore removed entirely.


# LSTM + Random Forest Hybrid Model

## Data Loading and Sequence Preparation

This notebook builds the hybrid LSTM + Random Forest model for credit card fraud detection. The same pre-processed engineered datasets from the EDA notebook are loaded, containing 1,296,675 training transactions and 555,719 test transactions across 14 features.

The key difference from the XGBoost standalone notebook is the addition of transaction sequencing. XGBoost treats every transaction independently — it reads one row at a time with no knowledge of what happened before. The LSTM component of the hybrid model requires the opposite: it reads a sequence of recent transactions from the same card in chronological order, looking for patterns in how spending behaviour changes over time.

To build these sequences, the original card numbers and timestamps are temporarily re-attached to the engineered features. Card numbers are used to group transactions by cardholder, and timestamps ensure the sequences are in the correct chronological order. These columns are used only for grouping and sorting — they are not fed into the model as features.

In [3]:
# Architecture diagram
from IPython.display import Image, display
import os

img_path = "hybrid_architecture.png"
if os.path.exists(img_path):
    display(Image(img_path, width=1000))
else:
    print("LSTM + Random Forest Hybrid Architecture:")
    print("  Input (14 features) -> Sequence Builder (5 timesteps) -> LSTM (64 units)")
    print("  -> Dropout (0.3) -> Dense (32) -> Dense (16) -> Fraud Probability")
    print("  Fraud Probability + 14 Static Features -> Random Forest -> Final Prediction")


LSTM + Random Forest Hybrid Architecture:
  Input (14 features) -> Sequence Builder (5 timesteps) -> LSTM (64 units)
  -> Dropout (0.3) -> Dense (32) -> Dense (16) -> Fraud Probability
  Fraud Probability + 14 Static Features -> Random Forest -> Final Prediction


In [4]:
# ============================================================
# CELL 4: BUILD TRANSACTION SEQUENCES FOR LSTM
# ============================================================
# Why sequences? XGBoost reads one transaction at a time.
# LSTM reads multiple transactions in ORDER — like reading
# a card's recent history. If a card had 5 purchases in
# the last hour, LSTM sees all 5 in order and spots the
# pattern. That's the whole point of the hybrid approach.
# ============================================================

from sklearn.preprocessing import StandardScaler
import numpy as np

# ----------------------------------------------------------
# Step 1: Identify which columns are actual features
# We added cc_num and trans_date_trans_time temporarily
# just for grouping and sorting. They are NOT model features.
# ----------------------------------------------------------
feature_cols = [c for c in X_train.columns if c not in ['cc_num', 'trans_date_trans_time']]
print(f"Features for model: {feature_cols}")
print(f"Number of features: {len(feature_cols)}")

# ----------------------------------------------------------
# Step 2: Scale all features to same range
# Why? LSTM is a neural network. Neural networks struggle
# when features have wildly different scales.
# amt ranges 0-10000, city_pop ranges 0-2000000, hour 0-23.
# StandardScaler converts everything to mean=0, std=1.
# fit_transform on train — learns the scaling rules
# transform on test — applies same rules (no data leakage)
# ----------------------------------------------------------
scaler = StandardScaler()
X_train[feature_cols] = scaler.fit_transform(X_train[feature_cols])
X_test[feature_cols] = scaler.transform(X_test[feature_cols])

# ----------------------------------------------------------
# Step 3: Sort by card number and time
# Why? LSTM reads transactions in chronological order.
# We need card 4521's transactions sorted oldest → newest.
# Without sorting, the sequences would be random order
# and LSTM can't learn any temporal patterns.
# We also temporarily attach the labels (y) so they stay
# aligned after sorting, then detach them again.
# ----------------------------------------------------------
X_train['y'] = y_train.values
X_train = X_train.sort_values(['cc_num', 'trans_date_trans_time']).reset_index(drop=True)
y_train = X_train['y']
X_train = X_train.drop(columns=['y'])

X_test['y'] = y_test.values
X_test = X_test.sort_values(['cc_num', 'trans_date_trans_time']).reset_index(drop=True)
y_test = X_test['y']
X_test = X_test.drop(columns=['y'])

# ----------------------------------------------------------
# Step 4: Define sequence length
# SEQ_LEN = 5 means for each transaction, LSTM looks at
# the LAST 5 transactions on that same card (including
# the current one). Like a sliding window.
#
# Example: Card 4521 has transactions T1, T2, T3, T4, T5, T6
# For T5, the sequence is [T1, T2, T3, T4, T5]
# For T6, the sequence is [T2, T3, T4, T5, T6]
#
# Why 5? Short enough to run fast on Colab.
# Long enough to capture rapid fraud patterns.
# ----------------------------------------------------------
SEQ_LEN = 5

# ----------------------------------------------------------
# Step 5: Function to build sequences per card
#
# For each transaction, this function:
# 1. Looks at the same card's previous transactions
# 2. Grabs the last SEQ_LEN transactions (the sequence)
# 3. If the card has fewer than SEQ_LEN transactions so far,
#    pads with zeros (e.g. card's 2nd ever transaction
#    gets [0, 0, 0, T1, T2] as its sequence)
# 4. Also saves the current transaction's static features
#    (these go to Random Forest, not LSTM)
# 5. Saves the label (fraud or legit) for that transaction
#
# Returns three arrays:
# - sequences: shape (num_transactions, SEQ_LEN, num_features)
#   This is what LSTM reads — 3D array
# - static_features: shape (num_transactions, num_features)
#   This is what Random Forest reads — 2D array
# - targets: shape (num_transactions,)
#   The fraud/legit labels
# ----------------------------------------------------------
def build_sequences(df, labels, feature_cols, seq_len):
    sequences = []        # Will hold all LSTM sequences
    static_features = []  # Will hold all RF static features
    targets = []          # Will hold all labels

    # Group all transactions by card number
    # Each group = one card's full transaction history
    grouped = df.groupby('cc_num')

    for card, group in grouped:
        # Get the feature values for this card's transactions
        # Already sorted by time from Step 3
        features = group[feature_cols].values

        # Get the labels for this card's transactions
        card_labels = labels.loc[group.index].values

        # Loop through each transaction on this card
        for i in range(len(group)):
            # Grab the last seq_len transactions up to and including current
            start = max(0, i - seq_len + 1)
            seq = features[start:i+1]

            # If this card doesn't have enough history yet, pad with zeros
            # E.g. card's 1st transaction: seq = [T1] → pad to [0,0,0,0,T1]
            # E.g. card's 3rd transaction: seq = [T1,T2,T3] → pad to [0,0,T1,T2,T3]
            if len(seq) < seq_len:
                padding = np.zeros((seq_len - len(seq), len(feature_cols)))
                seq = np.vstack([padding, seq])

            # Save the sequence (for LSTM)
            sequences.append(seq)

            # Save the current transaction features (for Random Forest)
            static_features.append(features[i])

            # Save the label
            targets.append(card_labels[i])

    return np.array(sequences), np.array(static_features), np.array(targets)

# ----------------------------------------------------------
# Step 6: Build the actual sequences
# This loops through all 1.29M training transactions
# grouped by 983 cards. Takes 3-5 minutes.
# ----------------------------------------------------------
print("Building training sequences...")
X_train_seq, X_train_static, y_train_seq = build_sequences(X_train, y_train, feature_cols, SEQ_LEN)
print(f"Train sequences: {X_train_seq.shape}")    # Expected: (1296675, 5, 14)
print(f"Train static: {X_train_static.shape}")      # Expected: (1296675, 14)
print(f"Train labels: {y_train_seq.shape} | Fraud: {int(y_train_seq.sum())}")

print("\nBuilding test sequences...")
X_test_seq, X_test_static, y_test_seq = build_sequences(X_test, y_test, feature_cols, SEQ_LEN)
print(f"Test sequences: {X_test_seq.shape}")        # Expected: (555719, 5, 14)
print(f"Test static: {X_test_static.shape}")          # Expected: (555719, 14)
print(f"Test labels: {y_test_seq.shape} | Fraud: {int(y_test_seq.sum())}")

Features for model: ['amt', 'city_pop', 'hour', 'month', 'distance_cardholder_merchant', 'age', 'is_weekend', 'is_night', 'velocity_1h', 'velocity_24h', 'amount_velocity_1h', 'category_encoded', 'gender_encoded', 'day_of_week_encoded']
Number of features: 14


Building training sequences...


Train sequences: (1296675, 5, 14)
Train static: (1296675, 14)
Train labels: (1296675,) | Fraud: 7506

Building test sequences...


Test sequences: (555719, 5, 14)
Test static: (555719, 14)
Test labels: (555719,) | Fraud: 2145


In [5]:
# CELL 5: LOAD PRE-TRAINED LSTM AND EXTRACT FEATURES
import zipfile, tempfile, os, joblib
import tensorflow as tf
from tensorflow.keras.layers import Input, LSTM, Dropout, Dense
from tensorflow.keras.models import Model
import numpy as np

print('TensorFlow version:', tf.__version__)

# Rebuild architecture (5 timesteps, 14 features)
inp = Input(shape=(5, 14), name='sequence_input')
x = LSTM(64, return_sequences=False)(inp)
x = Dropout(0.3)(x)
x = Dense(32, activation='relu')(x)
feat = Dense(16, activation='relu', name='lstm_features')(x)
out = Dense(1, activation='sigmoid')(feat)
train_model = Model(inputs=inp, outputs=out)

# Load saved weights from .keras archive
keras_path = os.path.join('..', 'models', 'saved', '02_comparator', 'lstm_rf_keras.keras')
with zipfile.ZipFile(keras_path, 'r') as z:
    z.extract('model.weights.h5', path=tempfile.gettempdir())
weights_path = os.path.join(tempfile.gettempdir(), 'model.weights.h5')
train_model.load_weights(weights_path)
print('Pre-trained LSTM weights loaded successfully')
train_model.summary()

# Feature-extraction sub-model
lstm_model = Model(inputs=inp, outputs=feat)
print('\nLSTM feature extractor output shape:', lstm_model.output_shape)

# Extract features from sequences
print('\nExtracting LSTM features from training data...')
train_lstm_features = lstm_model.predict(X_train_seq, batch_size=512, verbose=1)
print(f'LSTM train features: {train_lstm_features.shape}')

print('\nExtracting LSTM features from test data...')
test_lstm_features = lstm_model.predict(X_test_seq, batch_size=512, verbose=1)
print(f'LSTM test features: {test_lstm_features.shape}')

# Merge LSTM features + static features
X_train_merged = np.hstack([train_lstm_features, X_train_static])
X_test_merged = np.hstack([test_lstm_features, X_test_static])
print(f'\nMerged train: {X_train_merged.shape}')
print(f'Merged test: {X_test_merged.shape}')


TensorFlow version: 2.21.0


Pre-trained LSTM weights loaded successfully


Model: "functional"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ sequence_input (InputLayer)     │ (None, 5, 14)          │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ lstm (LSTM)                     │ (None, 64)             │        20,224 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout (Dropout)               │ (None, 64)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 32)             │         2,080 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ lstm_features (Dense)           │ (None, 16)             │           528 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 1)              │            17 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 22,849 (89.25 KB)

 Trainable params: 22,849 (89.25 KB)

 Non-trainable params: 0 (0.00 B)


LSTM feature extractor output shape: (None, 16)

Extracting LSTM features from training data...


   1/2533 ━━━━━━━━━━━━━━━━━━━━ 9:57 236ms/step

  10/2533 ━━━━━━━━━━━━━━━━━━━━ 15s 6ms/step   

  21/2533 ━━━━━━━━━━━━━━━━━━━━ 13s 5ms/step

  34/2533 ━━━━━━━━━━━━━━━━━━━━ 12s 5ms/step

  46/2533 ━━━━━━━━━━━━━━━━━━━━ 11s 5ms/step

  58/2533 ━━━━━━━━━━━━━━━━━━━━ 11s 5ms/step

  70/2533 ━━━━━━━━━━━━━━━━━━━━ 11s 5ms/step

  81/2533 ━━━━━━━━━━━━━━━━━━━━ 11s 5ms/step

  93/2533 ━━━━━━━━━━━━━━━━━━━━ 11s 5ms/step

 105/2533 ━━━━━━━━━━━━━━━━━━━━ 11s 5ms/step

 116/2533 ━━━━━━━━━━━━━━━━━━━━ 11s 5ms/step

 127/2533 ━━━━━━━━━━━━━━━━━━━━ 11s 5ms/step

 139/2533 ━━━━━━━━━━━━━━━━━━━━ 10s 5ms/step

 149/2533 ━━━━━━━━━━━━━━━━━━━━ 10s 5ms/step

 160/2533 ━━━━━━━━━━━━━━━━━━━━ 10s 5ms/step

 173/2533 ━━━━━━━━━━━━━━━━━━━━ 10s 5ms/step

 185/2533 ━━━━━━━━━━━━━━━━━━━━ 10s 5ms/step

 197/2533 ━━━━━━━━━━━━━━━━━━━━ 10s 5ms/step

 210/2533 ━━━━━━━━━━━━━━━━━━━━ 10s 5ms/step

 222/2533 ━━━━━━━━━━━━━━━━━━━━ 10s 4ms/step

 234/2533 ━━━━━━━━━━━━━━━━━━━━ 10s 4ms/step

 246/2533 ━━━━━━━━━━━━━━━━━━━━ 10s 4ms/step

 258/2533 ━━━━━━━━━━━━━━━━━━━━ 10s 4ms/step

 270/2533 ━━━━━━━━━━━━━━━━━━━━ 10s 4ms/step

 282/2533 ━━━━━━━━━━━━━━━━━━━━ 10s 4ms/step

 294/2533 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step 

 306/2533 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step

 319/2533 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step

 331/2533 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step

 343/2533 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step

 356/2533 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step

 369/2533 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step

 380/2533 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step

 390/2533 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step

 402/2533 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step

 414/2533 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step

 427/2533 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step

 439/2533 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step

 451/2533 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step

 463/2533 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step

 476/2533 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step

 489/2533 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step

 501/2533 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step

 513/2533 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step

 525/2533 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step

 537/2533 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step

 549/2533 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step

 561/2533 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step

 573/2533 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step

 585/2533 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step

 597/2533 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step

 609/2533 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step

 620/2533 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step

 632/2533 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step

 644/2533 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step

 656/2533 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step

 668/2533 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step

 680/2533 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step

 692/2533 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step

 704/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 716/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 729/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 742/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 754/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 766/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 779/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 791/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 803/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 815/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 827/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 840/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 851/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 863/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 875/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 887/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 898/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 910/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 922/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 934/2533 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step

 946/2533 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step

 958/2533 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step

 970/2533 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step

 982/2533 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step

 994/2533 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step

1006/2533 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step

1019/2533 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step

1031/2533 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step

1043/2533 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step

1056/2533 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step

1068/2533 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step

1079/2533 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step

1091/2533 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step

1104/2533 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step

1116/2533 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step

1128/2533 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step

1141/2533 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step

1153/2533 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step

1166/2533 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step

1178/2533 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step

1190/2533 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step

1202/2533 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step

1215/2533 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step

1228/2533 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step

1241/2533 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step

1254/2533 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step

1266/2533 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step

1278/2533 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step

1290/2533 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step

1302/2533 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step

1313/2533 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step

1325/2533 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step

1337/2533 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step

1349/2533 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step

1361/2533 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step

1373/2533 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step

1385/2533 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

1397/2533 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

1409/2533 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

1422/2533 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

1434/2533 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

1446/2533 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

1458/2533 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

1470/2533 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

1482/2533 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

1494/2533 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

1506/2533 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

1518/2533 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

1530/2533 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

1541/2533 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

1552/2533 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

1564/2533 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

1576/2533 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

1588/2533 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

1601/2533 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

1613/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1625/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1637/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1649/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1661/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1673/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1685/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1698/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1710/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1722/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1735/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1747/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1759/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1771/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1782/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1794/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1806/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1818/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1830/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1842/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

1854/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

1866/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

1878/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

1890/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

1903/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

1915/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

1927/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

1939/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

1951/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

1963/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

1975/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

1987/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

1999/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

2011/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

2023/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

2036/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

2048/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

2060/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

2073/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2085/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2097/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2109/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2121/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2133/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2145/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2157/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2170/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2182/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2194/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2206/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2218/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2231/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2242/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2254/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2266/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2278/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2290/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2302/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2313/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2325/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2337/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2350/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2363/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2375/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2388/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2401/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2413/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2426/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2438/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2450/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2462/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2473/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2485/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2497/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2509/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2521/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2533/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2533/2533 ━━━━━━━━━━━━━━━━━━━━ 11s 4ms/step


LSTM train features: (1296675, 16)

Extracting LSTM features from test data...
   1/1086 ━━━━━━━━━━━━━━━━━━━━ 44s 41ms/step

  12/1086 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step  

  22/1086 ━━━━━━━━━━━━━━━━━━━━ 5s 5ms/step

  34/1086 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step

  44/1086 ━━━━━━━━━━━━━━━━━━━━ 5s 5ms/step

  55/1086 ━━━━━━━━━━━━━━━━━━━━ 5s 5ms/step

  67/1086 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step

  80/1086 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step

  92/1086 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step

 104/1086 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step

 116/1086 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step

 128/1086 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step

 139/1086 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step

 152/1086 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step

 164/1086 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step

 176/1086 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step

 189/1086 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step

 201/1086 ━━━━━━━━━━━━━━━━━━━━ 3s 5ms/step

 213/1086 ━━━━━━━━━━━━━━━━━━━━ 3s 5ms/step

 225/1086 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

 237/1086 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

 249/1086 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

 261/1086 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

 271/1086 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

 282/1086 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

 294/1086 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

 306/1086 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

 318/1086 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

 330/1086 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

 343/1086 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

 355/1086 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

 367/1086 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

 379/1086 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

 391/1086 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

 402/1086 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

 414/1086 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

 426/1086 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

 438/1086 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

 450/1086 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

 462/1086 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

 474/1086 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

 486/1086 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

 496/1086 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

 507/1086 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

 519/1086 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

 531/1086 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

 543/1086 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

 555/1086 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

 567/1086 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

 579/1086 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

 591/1086 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

 603/1086 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

 615/1086 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

 627/1086 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

 639/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 651/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 663/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 675/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 687/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 700/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 712/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 722/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 733/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 745/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 757/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 769/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 781/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 793/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 805/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 817/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 828/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 839/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 851/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 863/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

 875/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

 886/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

 898/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

 910/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

 921/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

 933/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

 944/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

 955/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

 967/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

 980/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

 992/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

1004/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

1015/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

1027/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

1039/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

1051/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

1063/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

1075/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

1086/1086 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step


LSTM test features: (555719, 16)

Merged train: (1296675, 30)
Merged test: (555719, 30)


In [6]:
# CELL 6: TRAIN RF ON LSTM FEATURES + EVALUATE
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import (classification_report, f1_score, roc_auc_score,
                             precision_score, recall_score, accuracy_score,
                             average_precision_score)

# Get LSTM fraud probabilities
print('Getting LSTM fraud probabilities...')
train_lstm_prob = train_model.predict(X_train_seq, batch_size=512, verbose=0).flatten()
test_lstm_prob = train_model.predict(X_test_seq, batch_size=512, verbose=0).flatten()

# Build feature matrix: LSTM probability + 14 static features = 15 features
X_train_final = np.column_stack([train_lstm_prob, X_train_static])
X_test_final = np.column_stack([test_lstm_prob, X_test_static])
print(f'Feature matrix: {X_train_final.shape[1]} features')

# Train RF with class weights (fast - no hyperparameter search)
n_normal = (y_train_seq == 0).sum()
n_fraud = (y_train_seq == 1).sum()
scale_pos = n_normal / n_fraud

rf_model = RandomForestClassifier(
    n_estimators=200, max_depth=12,
    class_weight={0: 1, 1: scale_pos},
    random_state=42, n_jobs=-1
)
print(f'Training RF (class_weight ratio: {scale_pos:.1f})...')
rf_model.fit(X_train_final, y_train_seq)

# Predict
hybrid_pred = rf_model.predict(X_test_final)
hybrid_proba = rf_model.predict_proba(X_test_final)[:, 1]

print()
print("="*50)
print("HYBRID MODEL: LSTM + Random Forest")
print("="*50)
print(f"Accuracy:  {accuracy_score(y_test_seq, hybrid_pred):.4f}")
print(f"Precision: {precision_score(y_test_seq, hybrid_pred):.4f}")
print(f"Recall:    {recall_score(y_test_seq, hybrid_pred):.4f}")
print(f"F1 Score:  {f1_score(y_test_seq, hybrid_pred):.4f}")
print(f"ROC-AUC:   {roc_auc_score(y_test_seq, hybrid_proba):.4f}")
print(f"PR-AUC:    {average_precision_score(y_test_seq, hybrid_proba):.4f}")
print()
print("Classification Report:")
print(classification_report(y_test_seq, hybrid_pred, target_names=["Legit", "Fraud"]))


Getting LSTM fraud probabilities...


Feature matrix: 15 features
Training RF (class_weight ratio: 171.8)...



HYBRID MODEL: LSTM + Random Forest
Accuracy:  0.9978
Precision: 0.6522
Recall:    0.9487
F1 Score:  0.7730


ROC-AUC:   0.9980
PR-AUC:    0.9495

Classification Report:


              precision    recall  f1-score   support

       Legit       1.00      1.00      1.00    553574
       Fraud       0.65      0.95      0.77      2145

    accuracy                           1.00    555719
   macro avg       0.83      0.97      0.89    555719
weighted avg       1.00      1.00      1.00    555719



In [7]:
# CELL 7: SMOTE VARIANT
from imblearn.over_sampling import SMOTE
from sklearn.ensemble import RandomForestClassifier

smote = SMOTE(random_state=42)
X_train_smote, y_train_smote = smote.fit_resample(X_train_final, y_train_seq)
print(f'Before SMOTE: {dict(zip(*np.unique(y_train_seq, return_counts=True)))}')
print(f'After SMOTE:  {dict(zip(*np.unique(y_train_smote, return_counts=True)))}')

rf_smote = RandomForestClassifier(n_estimators=200, max_depth=10, random_state=42, n_jobs=-1)
print('Training RF + SMOTE...')
rf_smote.fit(X_train_smote, y_train_smote)

hybrid_smote_pred = rf_smote.predict(X_test_final)
hybrid_smote_proba = rf_smote.predict_proba(X_test_final)[:, 1]

print()
print('=' * 50)
print('HYBRID MODEL: LSTM + RF + SMOTE')
print('=' * 50)
print(f'Accuracy:  {accuracy_score(y_test_seq, hybrid_smote_pred):.4f}')
print(f'Precision: {precision_score(y_test_seq, hybrid_smote_pred):.4f}')
print(f'Recall:    {recall_score(y_test_seq, hybrid_smote_pred):.4f}')
print(f'F1 Score:  {f1_score(y_test_seq, hybrid_smote_pred):.4f}')
print(f'ROC-AUC:   {roc_auc_score(y_test_seq, hybrid_smote_proba):.4f}')
print()
print('Classification Report:')
print(classification_report(y_test_seq, hybrid_smote_pred, target_names=['Legit', 'Fraud']))


Before SMOTE: {np.int64(0): np.int64(1289169), np.int64(1): np.int64(7506)}
After SMOTE:  {np.int64(0): np.int64(1289169), np.int64(1): np.int64(1289169)}
Training RF + SMOTE...



HYBRID MODEL: LSTM + RF + SMOTE
Accuracy:  0.9953
Precision: 0.4499
Recall:    0.9650
F1 Score:  0.6137


ROC-AUC:   0.9968

Classification Report:
              precision    recall  f1-score   support

       Legit       1.00      1.00      1.00    553574
       Fraud       0.45      0.97      0.61      2145

    accuracy                           1.00    555719
   macro avg       0.72      0.98      0.81    555719
weighted avg       1.00      1.00      1.00    555719



In [8]:
# CELL 8: MODEL PARAMETERS
print('Saved RF model parameters:')
for k, v in rf_model.get_params().items():
    if k in ['n_estimators', 'max_depth', 'class_weight', 'min_samples_split', 'min_samples_leaf']:
        print(f'  {k}: {v}')
print(f'Final Hybrid F1 Score: {f1_score(y_test_seq, hybrid_pred):.4f}')


Saved RF model parameters:
  class_weight: {0: 1, 1: np.float64(171.75179856115108)}
  max_depth: 12
  min_samples_leaf: 1
  min_samples_split: 2
  n_estimators: 200
Final Hybrid F1 Score: 0.7730


In [9]:
# CELL 9: MODEL COMPARISON SUMMARY
print('=' * 65)
print('HYBRID MODEL VARIANTS COMPARISON')
print('=' * 65)

f1_saved = f1_score(y_test_seq, hybrid_pred)
p_saved = precision_score(y_test_seq, hybrid_pred)
r_saved = recall_score(y_test_seq, hybrid_pred)
auc_saved = roc_auc_score(y_test_seq, hybrid_proba)

f1_sm = f1_score(y_test_seq, hybrid_smote_pred)
p_sm = precision_score(y_test_seq, hybrid_smote_pred)
r_sm = recall_score(y_test_seq, hybrid_smote_pred)
auc_sm = roc_auc_score(y_test_seq, hybrid_smote_proba)

print(f'{"Variant":<30} {"F1":>8} {"Precision":>10} {"Recall":>8} {"ROC-AUC":>9}')
print('-' * 65)
print(f'{"LSTM + RF (class-weighted)":<30} {f1_saved:>8.4f} {p_saved:>10.4f} {r_saved:>8.4f} {auc_saved:>9.4f}')
print(f'{"LSTM + RF + SMOTE":<30} {f1_sm:>8.4f} {p_sm:>10.4f} {r_sm:>8.4f} {auc_sm:>9.4f}')
print('-' * 65)
print()
print('The class-weighted RF is the final selected hybrid variant.')


HYBRID MODEL VARIANTS COMPARISON


Variant                              F1  Precision   Recall   ROC-AUC
-----------------------------------------------------------------
LSTM + RF (class-weighted)       0.7730     0.6522   0.9487    0.9980
LSTM + RF + SMOTE                0.6137     0.4499   0.9650    0.9968
-----------------------------------------------------------------

The class-weighted RF is the final selected hybrid variant.


In [10]:
# Instead of 16 LSTM features, use LSTM's fraud prediction as 1 feature
train_lstm_prob = train_model.predict(X_train_seq, batch_size=512, verbose=1).flatten()
test_lstm_prob = train_model.predict(X_test_seq, batch_size=512, verbose=1).flatten()

# Merge: 1 LSTM probability + 14 static features = 15 features
X_train_slim = np.column_stack([train_lstm_prob, X_train_static])
X_test_slim = np.column_stack([test_lstm_prob, X_test_static])

print(f"Slim train: {X_train_slim.shape}")
print(f"Slim test: {X_test_slim.shape}")

# Train RF on slim features
rf_slim = RandomForestClassifier(
    n_estimators=200, max_depth=12,
    class_weight={0:1, 1:150},
    random_state=42, n_jobs=-1
)
rf_slim.fit(X_train_slim, y_train_seq)
slim_pred = rf_slim.predict(X_test_slim)
slim_proba = rf_slim.predict_proba(X_test_slim)[:, 1]

print("\n" + "="*50)
print("HYBRID v2: LSTM probability + RF")
print("="*50)
print(f"Accuracy:  {accuracy_score(y_test_seq, slim_pred):.4f}")
print(f"Precision: {precision_score(y_test_seq, slim_pred):.4f}")
print(f"Recall:    {recall_score(y_test_seq, slim_pred):.4f}")
print(f"F1 Score:  {f1_score(y_test_seq, slim_pred):.4f}")
print(f"ROC-AUC:   {roc_auc_score(y_test_seq, slim_proba):.4f}")
print("\nClassification Report:")
print(classification_report(y_test_seq, slim_pred, target_names=['Legit', 'Fraud']))

   1/2533 ━━━━━━━━━━━━━━━━━━━━ 2:09 51ms/step

  11/2533 ━━━━━━━━━━━━━━━━━━━━ 12s 5ms/step  

  23/2533 ━━━━━━━━━━━━━━━━━━━━ 12s 5ms/step

  35/2533 ━━━━━━━━━━━━━━━━━━━━ 11s 5ms/step

  47/2533 ━━━━━━━━━━━━━━━━━━━━ 11s 5ms/step

  59/2533 ━━━━━━━━━━━━━━━━━━━━ 11s 5ms/step

  71/2533 ━━━━━━━━━━━━━━━━━━━━ 10s 4ms/step

  83/2533 ━━━━━━━━━━━━━━━━━━━━ 10s 4ms/step

  96/2533 ━━━━━━━━━━━━━━━━━━━━ 10s 4ms/step

 108/2533 ━━━━━━━━━━━━━━━━━━━━ 10s 4ms/step

 120/2533 ━━━━━━━━━━━━━━━━━━━━ 10s 4ms/step

 133/2533 ━━━━━━━━━━━━━━━━━━━━ 10s 4ms/step

 144/2533 ━━━━━━━━━━━━━━━━━━━━ 10s 4ms/step

 155/2533 ━━━━━━━━━━━━━━━━━━━━ 10s 4ms/step

 167/2533 ━━━━━━━━━━━━━━━━━━━━ 10s 4ms/step

 179/2533 ━━━━━━━━━━━━━━━━━━━━ 10s 4ms/step

 191/2533 ━━━━━━━━━━━━━━━━━━━━ 10s 4ms/step

 204/2533 ━━━━━━━━━━━━━━━━━━━━ 10s 4ms/step

 216/2533 ━━━━━━━━━━━━━━━━━━━━ 10s 4ms/step

 229/2533 ━━━━━━━━━━━━━━━━━━━━ 10s 4ms/step

 242/2533 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step 

 254/2533 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step

 266/2533 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step

 278/2533 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step

 291/2533 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step

 304/2533 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step

 317/2533 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step

 330/2533 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step

 343/2533 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step

 357/2533 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step

 370/2533 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step

 381/2533 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step

 393/2533 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step

 405/2533 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step

 418/2533 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step

 430/2533 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step

 443/2533 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step

 456/2533 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step

 469/2533 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step

 482/2533 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step

 494/2533 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step

 507/2533 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step

 519/2533 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step

 531/2533 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step

 544/2533 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step

 556/2533 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step

 569/2533 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step

 582/2533 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step

 595/2533 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step

 607/2533 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step

 618/2533 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step

 629/2533 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step

 641/2533 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step

 654/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 667/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 679/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 692/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 704/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 717/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 729/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 741/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 754/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 766/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 778/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 791/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 804/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 817/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 830/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 843/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 854/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 866/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 878/2533 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step

 891/2533 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step

 904/2533 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step

 917/2533 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step

 929/2533 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step

 942/2533 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step

 955/2533 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step

 968/2533 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step

 981/2533 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step

 994/2533 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step

1006/2533 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step

1018/2533 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step

1031/2533 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step

1044/2533 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step

1057/2533 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step

1070/2533 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step

1083/2533 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step

1095/2533 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step

1106/2533 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step

1119/2533 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step

1131/2533 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step

1144/2533 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step

1157/2533 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step

1170/2533 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step

1182/2533 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step

1194/2533 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step

1208/2533 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step

1221/2533 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step

1233/2533 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step

1246/2533 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step

1258/2533 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step

1271/2533 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step

1285/2533 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step

1298/2533 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step

1311/2533 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step

1324/2533 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step

1336/2533 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step

1347/2533 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

1360/2533 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

1373/2533 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

1385/2533 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

1398/2533 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

1410/2533 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

1423/2533 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

1436/2533 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

1448/2533 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

1461/2533 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

1474/2533 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

1487/2533 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

1499/2533 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

1512/2533 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

1525/2533 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

1538/2533 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

1551/2533 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

1564/2533 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

1576/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1588/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1601/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1615/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1627/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1640/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1653/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1666/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1679/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1692/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1704/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1717/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1730/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1743/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1756/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1769/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1782/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1795/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1808/2533 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

1820/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

1831/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

1843/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

1856/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

1869/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

1882/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

1895/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

1908/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

1921/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

1934/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

1947/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

1959/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

1972/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

1984/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

1997/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

2010/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

2022/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

2035/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

2047/2533 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

2060/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2070/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2081/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2093/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2106/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2119/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2131/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2143/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2154/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2165/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2175/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2188/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2202/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2215/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2229/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2242/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2255/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2268/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2281/2533 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

2294/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2305/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2317/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2329/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2343/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2356/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2369/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2382/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2394/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2407/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2420/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2433/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2446/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2460/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2471/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2483/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2495/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2507/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2520/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2532/2533 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

2533/2533 ━━━━━━━━━━━━━━━━━━━━ 11s 4ms/step


   1/1086 ━━━━━━━━━━━━━━━━━━━━ 41s 38ms/step

  12/1086 ━━━━━━━━━━━━━━━━━━━━ 5s 5ms/step  

  24/1086 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step

  36/1086 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

  48/1086 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

  61/1086 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

  73/1086 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

  84/1086 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

  96/1086 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

 108/1086 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

 120/1086 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

 132/1086 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

 145/1086 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

 157/1086 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step

 171/1086 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

 183/1086 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

 195/1086 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

 206/1086 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

 217/1086 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

 230/1086 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

 242/1086 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

 254/1086 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

 266/1086 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

 278/1086 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

 291/1086 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

 304/1086 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

 316/1086 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

 328/1086 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

 341/1086 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

 354/1086 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

 366/1086 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

 378/1086 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

 389/1086 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step

 400/1086 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

 412/1086 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

 424/1086 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

 434/1086 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

 446/1086 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

 459/1086 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

 472/1086 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

 485/1086 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

 498/1086 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

 510/1086 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

 523/1086 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

 535/1086 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

 548/1086 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

 560/1086 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

 574/1086 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

 587/1086 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

 600/1086 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

 612/1086 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step

 625/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 637/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 650/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 662/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 673/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 685/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 698/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 711/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 724/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 737/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 750/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 762/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 774/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 786/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 798/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 811/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 824/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 837/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 850/1086 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step

 863/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

 875/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

 887/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

 899/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

 910/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

 922/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

 934/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

 947/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

 960/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

 973/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

 985/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

 998/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

1009/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

1021/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

1033/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

1046/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

1059/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

1072/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

1084/1086 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step

1086/1086 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step


Slim train: (1296675, 15)
Slim test: (555719, 15)



HYBRID v2: LSTM probability + RF
Accuracy:  0.9980
Precision: 0.6685
Recall:    0.9469
F1 Score:  0.7837


ROC-AUC:   0.9981

Classification Report:
              precision    recall  f1-score   support

       Legit       1.00      1.00      1.00    553574
       Fraud       0.67      0.95      0.78      2145

    accuracy                           1.00    555719
   macro avg       0.83      0.97      0.89    555719
weighted avg       1.00      1.00      1.00    555719



## Final result summary — Hybrid 1 (LSTM + RF)

| Pipeline | F1 | Precision | Recall | ROC-AUC | Source |
|---|---:|---:|---:|---:|---|
| **Corrected (composite-key merge)** | **0.7892** | 0.6770 | 0.9459 | 0.9981 | `verified_metrics.json` → `LSTM_reproduced_baseline` |
| Historical / broken (positional `.values` attach) | 0.4747 | — | — | — | `FYP_Hybrid_Model_BROKEN.ipynb` (audit-preserved only) |

The 0.4747 figure is **not** the current Hybrid 1 result and should not be cited as such in the dissertation. It is preserved purely for transparency about the methodology fix.

Re-run the corrected pipeline end-to-end from the CLI with:

```
python run_gap_experiments.py
```

That script uses the same `_with_ids.csv` files and seed=42 and writes the authoritative metrics block to `verified_metrics.json`.


# Final 3-Model Staged Study Summary

This notebook implements the **Hybrid Comparator** of the staged 3-model study. Its role is to test sequence/temporal modelling against the tabular Baseline and Proposed Model. The headline 3-model comparison (and the Proposed Model component analysis) is also recorded in `04_BDS_GA.ipynb`.

## Main Model Comparison

| Category | Model | F1 | Precision | Recall | ROC-AUC | PR-AUC |
|---|---|---|---|---|---|---|
| Baseline | XGBoost (SMOTE+tuned) | 0.8646 | 0.9297 | 0.8079 | 0.9972 | 0.9092 |
| Hybrid Comparator | LSTM + Random Forest | 0.7892 | 0.6770 | 0.9459 | 0.9981 | n/a (not recorded) |
| Proposed Model | AE + BDS(GA) + XGBoost | 0.8706 | 0.9338 | 0.8154 | 0.9976 | 0.9158 |

## Proposed Model Component Analysis

| Variant | What It Tests | F1 | Precision | Recall | ROC-AUC | PR-AUC |
|---|---|---|---|---|---|---|
| XGBoost only | Without anomaly/BDS | 0.8646 | 0.9297 | 0.8079 | 0.9972 | 0.9092 |
| AE + XGBoost | Adds autoencoder | 0.8690 | 0.9369 | 0.8103 | 0.9973 | 0.9142 |
| AE + BDS(GA) + XGBoost | Adds behavioural scores | 0.8706 | 0.9338 | 0.8154 | 0.9976 | 0.9158 |

LSTM result here reflects post row-alignment fix; the historical 0.4747 figure is audit-preserved in `notebooks/archive/FYP_Hybrid_Model_BROKEN.ipynb` and is not for citation. All metrics at threshold = 0.5, sourced from `results/verified_metrics.json`.